source/extensions/filters/listener/http_inspector/http_inspector.cc

```diff
@@ -132,31 +155,26 @@ ParseState Filter::parseHttpHeader(absl::string_view data) {
     protocol_ = "HTTP/2";
     return ParseState::Done;
   } else {
-    const size_t pos = data.find_first_of("\r\n");
-    if (pos != absl::string_view::npos) {
-      const absl::string_view request_line = data.substr(0, pos);
-      const std::vector<absl::string_view> fields =
-          absl::StrSplit(request_line, absl::MaxSplits(' ', 4));
-
-      // Method SP Request-URI SP HTTP-Version
-      if (fields.size() != 3) {
-        ENVOY_LOG(trace, "http inspector: invalid http1x request line");
-        // done(false);
-        return ParseState::Error;
-      }
+    ssize_t rc =
+        http_parser_execute(&parser_, &settings_, data.data() + read_, data.length() - read_);
+    ENVOY_LOG(trace,
+              "http inspector: http_parser parsed {} chars, error code: {}, parser state: {}, "
+              "message complete: {}",
+              rc, HTTP_PARSER_ERRNO(&parser_), parser_.state, message_complete_);
+
+    // Errors in parsing HTTP.
+    if (HTTP_PARSER_ERRNO(&parser_) != HPE_OK && HTTP_PARSER_ERRNO(&parser_) != HPE_PAUSED) {
+      return ParseState::Error;
+    }
 
-      if (http1xMethods().count(fields[0]) == 0 || httpProtocols().count(fields[2]) == 0) {
-        ENVOY_LOG(trace, "http inspector: method: {} or protocol: {} not valid", fields[0],
-                  fields[2]);
-        // done(false);
-        return ParseState::Error;
+    // When the header is parsed, the http_parser will reset the state to s_start_req.
+    if (message_complete_) {
```

Comment:
Hmm, if you're trying to parse full HTTP request message, I don't think this work for large HTTP request with body, can you test with large header (> `MAX_INSPECT_SIZE`) or large request body (> `MAX_INSPECT_SIZE`)?

Reply:
Yeah, it won't work. I forgot the `MAX_INSPECT_SIZE`. I will add a text case to cover `MAX_INSPECT_SIZE`. But this raises a new problem, the HTTP inspector cannot guarantee to have the same output as HTTP1 codec for large header or large request body. Do you think we can just verify the HTTP headers without caring about the request body? If the request header set the content-length, should we verify the request body?

Reply:
I don't think you should do anything over request line, and that's how we determined `MAX_INSPECT_SIZE` to be 8k https://github.com/envoyproxy/envoy/pull/7559#discussion_r305173195

Anything beyond there should be handled by HCM, and I don't think inspector should interfere there, it also gives chance to HCM to return protocol error to client: https://github.com/envoyproxy/envoy/blob/75b78d8134315c872099837d4abf56aa7fe53d09/source/common/http/http1/codec_impl.cc#L728 

HCM does more stuff based on configuration such as `max_headers_kb`, `max_headers_count` etc, I don't think we want that in http_inspector either.

Reply:
Update the PR to validate the request line only